src/self_driving/planning/pilot.cpp

```diff
@@ -32,27 +50,355 @@ Pilot::Pilot(std::string model_save_path, std::string forecast_model_save_path,
+    for (size_t i = 0; i < json_decomp.size(); i++) {
+      cves.emplace_back(parser::ConstantValueExpression::FromString(json_decomp[i], types[i]));
+    }
+    return cves;
+  };
+
+  for (auto &info : out_metadata) {
+    metadata.query_id_to_dboid_[info.first] = info.second.db_oid_.UnderlyingValue();
+    metadata.query_id_to_text_[info.first] = info.second.text_.substr(1, info.second.text_.size() - 2);
+    metadata.query_id_to_param_types_[info.first] = types_conv(info.second.param_type_);
+  }
+
+  for (auto &info : out_params) {
+    for (auto &cve : info.second) {
+      metadata.query_id_to_params_[info.first].emplace_back(cves_conv(metadata, info.first, cve));
+    }
+  }
+
+  bool result = true;
+  execution::exec::ExecutionSettings settings{};
+  query_exec_util_->BeginTransaction(catalog::INVALID_DATABASE_OID);
+  {
+    // Metadata query
+    auto to_row_fn = [&metadata, types_conv](const std::vector<execution::sql::Val *> &values) {
+      auto db_oid = static_cast<execution::sql::Integer *>(values[0])->val_;
+      auto qid = execution::query_id_t(static_cast<execution::sql::Integer *>(values[1])->val_);
+
+      // Only insert new if not convered already
+      if (metadata.query_id_to_dboid_.find(qid) == metadata.query_id_to_dboid_.end()) {
+        metadata.query_id_to_dboid_[qid] = db_oid;
+
+        auto *text_val = static_cast<execution::sql::StringVal *>(values[2]);
+        // We do this since the string has been quoted by the metric
+        metadata.query_id_to_text_[qid] =
+            std::string(text_val->StringView().data() + 1, text_val->StringView().size() - 2);
+
+        auto *param_types = static_cast<execution::sql::StringVal *>(values[3]);
+        metadata.query_id_to_param_types_[qid] = types_conv(std::string(param_types->StringView()));
+      }
+    };
+
+    // This loads the entire query text history from the internal tables. It might be possible to
+    // do on-demand fetching or windowed fetching at a futrure time. We do this because a interval
+    // can execute a prepared query without a corresponding text recording (if the query was
+    // already prepared during a prior interval).
+    auto query = "SELECT * FROM noisepage_forecast_texts";
+    result &= query_exec_util_->ExecuteDML(query, nullptr, nullptr, to_row_fn, nullptr,
+                                           std::make_unique<optimizer::TrivialCostModel>(), settings);
+  }
+
+  {
+    auto to_row_fn = [&metadata, cves_conv](const std::vector<execution::sql::Val *> &values) {
+      auto qid = execution::query_id_t(static_cast<execution::sql::Integer *>(values[1])->val_);
+      auto *param_val = static_cast<execution::sql::StringVal *>(values[2]);
+      {
+        // Read the parameters. In the worse case, we will have double the parameters, but that is
+        // okay since every parameter will be duplicated. This can happen since the parameters
+        // could already be visible by the time this select query runs.
```

Comment:
Why not do a duplication-checking here as well, just like above?

Reply:
In the modified version, parameters are still logged whenever we are going to perform planning (i.e., parameters are logged at a batch timestamp). However, I've modified the range scan query to exclude the most recent batch of parameters written.